File: parser/parser.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/xeipuuv/gojsonschema"
)

type AttributeConfig struct {
	Field     string      `json:"field"`
	Source    string      `json:"source"`
	Required  bool        `json:"required"`
	Mapping   string      `json:"mapping"`
	SkipEmpty bool        `json:"skip_empty"`
	Default   interface{} `json:"default"`
}

var generatorConfig map[string]map[string][]AttributeConfig

func loadGeneratorConfig() error {
	configData, err := os.ReadFile("generator_config.json")
	if err != nil {
		return fmt.Errorf("error loading generator config: %w", err)
	}

	if err := json.Unmarshal(configData, &generatorConfig); err != nil {
		return fmt.Errorf("error parsing generator config: %w", err)
	}

	return nil
}

func loadSchema(schemaPath string) (*gojsonschema.Schema, error) {
	absPath, err := filepath.Abs(schemaPath)
	if err != nil {
		return nil, fmt.Errorf("error getting absolute path: %w", err)
	}

	schemaLoader := gojsonschema.NewReferenceLoader("file://" + absPath)
	schema, err := gojsonschema.NewSchema(schemaLoader)
	if err != nil {
		return nil, fmt.Errorf("error loading schema: %w", err)
	}

	return schema, nil
}

func validateConfig(schema *gojsonschema.Schema, config map[string]interface{}) (bool, string) {
	configJSON, err := json.Marshal(config)
	if err != nil {
		return false, fmt.Sprintf("error marshaling config: %v", err)
	}

	documentLoader := gojsonschema.NewBytesLoader(configJSON)
	result, err := schema.Validate(documentLoader)
	if err != nil {
		return false, fmt.Sprintf("validation error: %v", err)
	}

	if !result.Valid() {
		var errors []string
		for _, desc := range result.Errors() {
			errors = append(errors, desc.String())
		}
		return false, strings.Join(errors, "; ")
	}

	provider, _ := config["provider"].(string)

	// provider-specific validation
	if provider == "gcp" { // GCP
		services, _ := config["services"].([]interface{})
		for _, svc := range services {
			service, ok := svc.(map[string]interface{})
			if !ok {
				continue
			}
			if _, hasProjectID := service["project_id"]; !hasProjectID {
				return false, "GCP compute.instance requires 'project_id' in service configuration"
			}
		}
	} else if provider == "azure" { // Azure
		services, _ := config["services"].([]interface{})
		for _, svc := range services {
			service, ok := svc.(map[string]interface{})
			if !ok {
				continue
			}
			if _, hasSubId := service["subscription_id"]; !hasSubId {
				return false, "GCP compute.instance requires 'subscription_id' in service configuration"
			}
		}
	}

	return true, ""
}

func formatTfvarsValue(value interface{}) string {
	switch v := value.(type) {
	case string:
		return fmt.Sprintf(`"%s"`, v)
	case float64:
		if v == float64(int64(v)) {
			return fmt.Sprintf("%d", int64(v))
		}
		return fmt.Sprintf("%g", v)
	case map[string]interface{}:
		var items []string
		for k, val := range v {
			items = append(items, fmt.Sprintf("  %s = %s", k, formatTfvarsValue(val)))
		}
		return "{\n" + strings.Join(items, "\n") + "\n}"
	case []interface{}:
		var items []string
		for _, item := range v {
			items = append(items, formatTfvarsValue(item))
		}
		return "[" + strings.Join(items, ", ") + "]"
	default:
		return fmt.Sprintf("%v", v)
	}
}

func generateTfvars(provider, serviceType string, config, service map[string]interface{}) string {
	var lines []string

	// Get attribute configuration based on provider & service
	providerConfig, ok := generatorConfig[provider]
	if !ok {
		return ""
	}

	attrs, ok := providerConfig[serviceType]
	if !ok {
		return ""
	}

	for _, attr := range attrs {
		// Determine source map
		sourceMap := service
		if attr.Source == "config" {
			sourceMap = config
		}

		// Get source field name (use mapping if specified, otherwise use field name)
		sourceField := attr.Field
		if attr.Mapping != "" {
			sourceField = attr.Mapping
		}

		value, exists := sourceMap[sourceField]

		// Handle default values
		if !exists && attr.Default != nil {
			value = attr.Default
			exists = true
		}

		// Skip if not exists
		if !exists {
			continue
		}

		// Skip empty values if specified
		if attr.SkipEmpty {
			if strVal, ok := value.(string); ok && strVal == "" {
				continue
			}
			if mapVal, ok := value.(map[string]interface{}); ok && len(mapVal) == 0 {
				continue
			}
		}

		// Format and append the line
		formattedValue := formatTfvarsValue(value)
		lines = append(lines, fmt.Sprintf("%s = %s", attr.Field, formattedValue))
	}

	return strings.Join(lines, "\n") + "\n"
}

func getServiceFolderName(serviceType string) string {
	switch serviceType {
	case "compute.instance":
		return "compute_instance"
	case "storage.object":
		return "storage_object"
	}
	return strings.ReplaceAll(serviceType, ".", "_")
}

func parse(configPath string) bool {
	// Load configuration
	configData, err := os.ReadFile(configPath)
	if err != nil {
		fmt.Printf("[ERROR] Error loading configuration: %v\n", err)
		return false
	}

	var config map[string]interface{}
	if err := json.Unmarshal(configData, &config); err != nil {
		fmt.Printf("[ERROR] Error parsing configuration JSON: %v\n", err)
		return false
	}

	// Load and validate schema
	schema, err := loadSchema("schema.json")
	if err != nil {
		fmt.Printf("[ERROR] %v\n", err)
		return false
	}

	isValid, errorMsg := validateConfig(schema, config)
	if !isValid {
		fmt.Printf("[ERROR] Validation failed: %s\n", errorMsg)
		return false
	}

	fmt.Println("✓ Configuration validated successfully")

	// Generate version dir based on timestamp
	versionDir := fmt.Sprintf("tfvars_%s", time.Now().Format("20060102_150405"))

	provider, _ := config["provider"].(string)
	services, _ := config["services"].([]interface{})

	outputDir := filepath.Join("..", "deployment", provider, versionDir)

	// Process each service
	for _, svc := range services {
		service, ok := svc.(map[string]interface{})
		if !ok {
			continue
		}

		if err := os.MkdirAll(outputDir, 0755); err != nil {
			fmt.Printf("[ERROR] Error creating directory %s: %v\n", outputDir, err)
			return false
		}

		// Generate .tfvars content based on provider
		serviceType, _ := service["type"].(string)
		tfvarsContent := generateTfvars(provider, serviceType, config, service)

		// Name .tfvars file based on service
		tfvarsFile := filepath.Join(outputDir, fmt.Sprintf("%s.tfvars", getServiceFolderName(service["type"].(string))))
		if err := os.WriteFile(tfvarsFile, []byte(tfvarsContent), 0644); err != nil {
			fmt.Printf("[ERROR] Error writing file %s: %v\n", tfvarsFile, err)
			return false
		}

		fmt.Printf("✓ Generated: %s\n", tfvarsFile)
	}

	fmt.Printf("\n✓ Successfully parsed configuration\n")
	fmt.Printf("  Output directory: %s\n", outputDir)
	return true
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Usage: go run parser.go <config.json>")
		os.Exit(1)
	}

	// Load generator configuration
	if err := loadGeneratorConfig(); err != nil {
		fmt.Printf("[ERROR] %v\n", err)
		os.Exit(1)
	}

	configPath := os.Args[1]

	if !parse(configPath) {
		os.Exit(1)
	}
}

File: parser/parser.py
#!/usr/bin/env python3
"""
Multi-Cloud Infrastructure Deployment Parser

This script parses JSON deployment configurations and generates .tfvars files
for OpenTofu/Terraform deployments across AWS, GCP, and Azure.
"""

import json
import os
import sys
from datetime import datetime
from pathlib import Path
from typing import Dict, List, Any, Optional
import jsonschema


class DeploymentParser:
    def __init__(self, schema_path: str):
        self.schema = self._load_schema(schema_path=Path(schema_path))

    def _load_schema(self, schema_path) -> Dict[str, Any]:
        try:
            with open(schema_path, 'r') as f:
                return json.load(f)
        except Exception as e:
            print(f"Error loading schema: {e}")
            sys.exit(1)

    def validate_config(self, config: Dict[str, Any]) -> tuple:
        try:
            jsonschema.validate(instance=config, schema=self.schema)
        except Exception as e:
            return False, str(e)

        # Specific validation for different providers
        provider = config.get("provider")
        services = config.get("services", [])

        # Validate specific GCP
        if provider == "gcp":
            for service in services:
                if service.get("type") == "compute.instance" and not service.get("project_id"):
                    return False, "GCP compute.instance requires 'project_id' in service configuration"
        return True, ""

    def _generate_version_dir(self, config: Dict[str, Any]) -> Path:
        # Version set in config or timestamp as default
        version = config.get("version")
        if not version:
            timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
            version = f"deployment_{timestamp}"
        return Path(version)

    def _format_tfvars_value(self, value: Any) -> str:
        # Format a value for .tfvars file
        if isinstance(value, str):
            return f'"{value}"'
        elif isinstance(value, (int, float)):
            return str(value)
        elif isinstance(value, bool):
            return "true" if value else "false"
        elif isinstance(value, dict):
            items = [f'  {k} = {self._format_tfvars_value(v)}' for k, v in value.items()]
            return "{\n" + "\n".join(items) + "\n}"
        elif isinstance(value, list):
            items = [self._format_tfvars_value(item) for item in value]
            return "[" + ", ".join(items) + "]"
        else:
            return str(value)

    def _generate_aws_tfvars(self, config: Dict[str, Any], service: Dict[str, Any]) -> str:
        # Generate AWS .tfvars content
        lines = []
        lines.append(f'region         = "{config["region"]}"')
        lines.append(f'instance_id    = "{service["instance_id"]}"')
        lines.append(f'size           = "{service["size"]}"')
        lines.append(f'os             = "{service["os"]}"')

        if "disk_size_gb" in service:
            lines.append(f'disk_size_gb   = {service["disk_size_gb"]}')

        if "metadata" in service and service["metadata"]:
            metadata_str = self._format_tfvars_value(service["metadata"])
            lines.append(f'metadata = {metadata_str}')

        if "ssh_public_key" in service:
            lines.append(f'ssh_public_key = "{service.get("ssh_public_key", "")}"')
        else:
            lines.append('ssh_public_key = ""')

        return "\n".join(lines) + "\n"

    def _generate_gcp_tfvars(self, config: Dict[str, Any], service: Dict[str, Any]) -> str:
        # Generate GCP .tfvars content
        lines = []
        if "project_id" in service:
            lines.append(f'project_id  = "{service["project_id"]}"')
        lines.append(f'region      = "{config["region"]}"')
        lines.append(f'instance_id = "{service["instance_id"]}"')
        lines.append(f'size        = "{service["size"]}"')
        lines.append(f'os          = "{service["os"]}"')

        if "disk_size_gb" in service:
            lines.append(f'disk_size_gb = {service["disk_size_gb"]}')

        if "metadata" in service and service["metadata"]:
            metadata_str = self._format_tfvars_value(service["metadata"])
            lines.append(f'metadata = {metadata_str}')

        return "\n".join(lines) + "\n"

    def _generate_azure_tfvars(self, config: Dict[str, Any], service: Dict[str, Any]) -> str:
        # Generate Azure .tfvars content
        lines = []
        # Azure uses "location" instead of "region"
        lines.append(f'location    = "{config["region"]}"')
        lines.append(f'instance_id = "{service["instance_id"]}"')
        lines.append(f'size        = "{service["size"]}"')
        lines.append(f'os          = "{service["os"]}"')

        if "disk_size_gb" in service:
            lines.append(f'disk_size_gb = {service["disk_size_gb"]}')

        if "metadata" in service and service["metadata"]:
            metadata_str = self._format_tfvars_value(service["metadata"])
            lines.append(f'metadata = {metadata_str}')

        if "admin_username" in service:
            lines.append(f'admin_username = "{service["admin_username"]}"')

        if "ssh_public_key" in service:
            lines.append(f'ssh_public_key = "{service.get("ssh_public_key", "")}"')

        return "\n".join(lines) + "\n"

    def _get_service_folder_name(self, service_type: str) -> str:
        # Get folder name for service type
        if service_type == "compute.instance":
            return "compute_instance"
        elif service_type == "storage.object":
            return "storage_object"
        else:
            return service_type.replace(".", "_")

    def parse(self, config_path: str) -> bool:
        # Load configuration
        try:
            with open(config_path, 'r') as f:
                config = json.load(f)
        except Exception as e:
            print(f"[ERROR] Error loading configuration: {e}")
            return False

        # Validate configuration
        is_valid, error_msg = self.validate_config(config)
        if not is_valid:
            print(f"[ERROR] Validation failed: {error_msg}")
            return False

        print(f"✓ Configuration validated successfully")

        # Generate version directory
        version_dir = self._generate_version_dir(config)
        provider = config["provider"]
        services = config["services"]

        # Create output directory structure
        output_dir = Path("../opentofu") / provider / version_dir

        # Process each service
        for service in services:
            service_type = service["type"]
            service_folder = self._get_service_folder_name(service_type)

            # Create service directory
            service_dir = output_dir / service_folder
            service_dir.mkdir(parents=True, exist_ok=True)

            # Generate .tfvars content based on provider
            if provider == "aws":
                tfvars_content = self._generate_aws_tfvars(config, service)
            elif provider == "gcp":
                tfvars_content = self._generate_gcp_tfvars(config, service)
            elif provider == "azure":
                tfvars_content = self._generate_azure_tfvars(config, service)
            else:
                print(f"[ERROR] Unknown provider {provider}")
                return False

            # Write .tfvars file
            tfvars_file = service_dir / f'opentofu.tfvars'
            with open(tfvars_file, 'w') as f:
                f.write(tfvars_content)

            print(f"✓ Generated: {tfvars_file}")

        print(f"\n✓ Successfully parsed configuration")
        print(f"  Output directory: {output_dir}")
        return True

SCHEMA_PATH = Path(__file__).parent / "schema.json"
def main():
    """Main entry point."""
    if len(sys.argv) < 1:
        print("Usage: python parser.py <config.json>")
        sys.exit(1)

    config_path = sys.argv[1]

    # Get the directory where the script is located
    parser = DeploymentParser(schema_path=str(SCHEMA_PATH))
    success = parser.parse(config_path)
    
    return success


if __name__ == "__main__":
    main()
